lib/ddtrace/profiling/ext/cpu.rb

```diff
@@ -19,6 +20,65 @@ def self.apply!
           # will provide a thread/clock ID for CPU timing.
           require 'ddtrace/profiling/ext/cthread'
           ::Thread.send(:prepend, Profiling::Ext::CThread)
+
+          # Applying hooks will allow any existing threads to update
+          # their thread/clock IDs so their CPU timings can be measured.
+          apply_hooks!
+        end
+
+        def self.apply_hooks!
+          # Threads that have already been created, will not have resolved
+          # a thread/clock ID. This is because these IDs can only be resolved
+          # from within the thread's execution context, which we do not control.
+          #
+          # We can work around this by applying trace hooks to capture each thread
+          # from within its own execution context, resolving the thread/clock IDs,
+          # then disabling the hook so it doesn't re-run needlessly.
+          Thread.list.each do |thread|
+            # Only attempt to hook into threads that are instrumented
+            next unless thread.is_a?(CThread)
+
+            # TracePoint supports filtering by thread for Ruby 2.7+
+            if Gem::Version.new(RUBY_VERSION) >= Gem::Version.new('2.7') && defined?(TracePoint)
+              thread[THREAD_HOOK] = TracePoint.new(:line) do |tp|
+                current_thread = Thread.current
+
+                begin
+                  current_thread.send(:update_native_ids)
+                rescue StandardError => e
+                  log "ERROR: Failed to update thread/clock IDs for Thread #{current_thread.object_id}.\nCause: #{e.message}\nLocation: #{e.backtrace.first}"
+                ensure
+                  current_thread[THREAD_HOOK].disable
```

Comment:
A few ideas on how we could make this even safer:
1. Move the disabling of the tracepoint to the absolute top of the tracepoint block.
2. Use the `tp` reference to disable the tracepoint -- `tp.disable`. 

  This way we are absolutely sure we disable this tracepoint, even if for some weird reason `current_thread[THREAD_HOOK]` is not set correctly or is overwritten by a bug

Reply:
I like this; will do.